torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -0,0 +1,268 @@
+    edge_list edges;
+    // Create some dummy edges (input_nr not important for init_to_execute).
+    for (const auto& mapEntry : sendFunctions) {
+      edges.emplace_back(mapEntry.second, 0);
+    }
+
+    // Add the original graphRoot as an edge.
+    edges.emplace_back(graphRoot, 0);
+
+    // Create a dummy GraphRoot and run init_to_execute with it.
+    GraphRoot dummyRoot(edges, {});
+    graphTask->init_to_execute(dummyRoot, outputEdges);
+
+    // Mark all 'RecvRPCBackward' as needing execution.
+    for (const auto& recvBackwardEdge : recvBackwardEdges) {
+      graphTask->exec_info_[recvBackwardEdge.function.get()].needed_ = true;
+    }
+  }
+
+  // Let autograd context take ownership of the GraphTask.
+  autogradContext.setGraphTask(std::move(graphTask));
+}
+
+void DistEngine::runEngineAndAccumulateGradients(
+    DistAutogradContext& autogradContext,
+    std::shared_ptr<Node> graphRoot,
+    const edge_list& outputEdges) {
+  // Kick off autograd computation with the root node and retrieve all the
+  // gradients.
+  // TODO: make this non-blocking
+  // (https://github.com/pytorch/pytorch/issues/26359)
+  variable_list grads = engine_.execute_with_graph_task(
+      autogradContext.retrieveGraphTask(), graphRoot);
+
+  // Accumulate all the gradients in the context.
+  TORCH_INTERNAL_ASSERT(grads.size() == outputEdges.size());
+  for (size_t i = 0; i < grads.size(); i++) {
+    // It is possible that the grad is not defined since a separate invocation
+    // of the autograd engine on the same node might actually compute this
+    // gradient.
+    // Also accumulate grads only for AccumulateGrad function.
+    if (grads[i].defined() &&
+        dynamic_cast<AccumulateGrad*>(outputEdges[i].function.get())) {
+      auto& variable =
+          std::static_pointer_cast<AccumulateGrad>(outputEdges[i].function)
+              ->variable;
+      autogradContext.accumulateGrad(variable, grads[i]);
+    }
+  }
+}
+
+void DistEngine::executeSendFunction(
+    DistAutogradContext& autogradContext,
+    std::shared_ptr<Node> sendFunction) {
+  std::unique_lock<std::mutex> lock(initializedContextIdsLock_);
+  if (initializedContextIds_.find(autogradContext.contextId()) ==
+      initializedContextIds_.end()) {
+    edge_list outputEdges;
+    computeDependencies(autogradContext, {}, {}, sendFunction, outputEdges);
+
+    // Mark the autograd context id as initialized and unlock.
+    initializedContextIds_.insert(autogradContext.contextId());
+    lock.unlock();
+
+    runEngineAndAccumulateGradients(autogradContext, sendFunction, outputEdges);
+
+    // Wait for all of the outstanding rpcs to complete.
+    autogradContext.waitForOutStandingRpcs();
+  } else {
+    lock.unlock();
+    auto& graphTask = autogradContext.retrieveGraphTask();
+    engine_.enqueue_on_cpu(torch::autograd::NodeTask(
+        &graphTask, sendFunction, torch::autograd::InputBuffer(0)));
+  }
+}
+
+void DistEngine::execute(const variable_list& roots) {
```

Comment:
No support for explicit gradients in distributed backwards? I guess it can always be added later, but this might be a code smell (too much code duplicated, more should have been shared between the implementations.)

Reply:
Another smell: you don't clear callbacks in this codepath. Callbacks are used by DistributedDataParallel to run some custom code at the end of backwards. It seems orthogonal to whether or not you are distributed or not.

Reply:
> Regarding callbacks, isn't that a property of the local autograd engine? Ideally, users should only use the distributed autograd APIs and shouldn't mix this with APIs for the local autograd engine. Although, I'm not sure what is the right way of enforcing this.

If you're not going to handle callbacks correctly you should make it so that e.g. `queue_callbacks` raises an error. But I fell like it is *really* simple to handle all of this correctly and reduce code duplication...

Reply:
Looking into the callbacks a bit more, wouldn't the callbacks be cleared by the local autograd engine? This method calls the local autograd engine underneath which will setup the ClearCallbacks guard, run all the callbacks at the end and finally clear the callbacks when the guard dies?